pool-apps/pool/src/lib/error.rs

```diff
@@ -20,11 +20,75 @@ use stratum_apps::{
         noise_sv2,
         parsers_sv2::{Mining, ParserError},
     },
-    utils::types::ChannelId,
+    utils::types::{ChannelId, DownstreamId},
 };
 
 pub type PoolResult<T> = Result<T, PoolError>;
 
+#[derive(Debug)]
+pub enum SourceCtx {
```

Comment:
I don't see why we should abbreviate words here, we can write the entire word.

```suggestion
pub enum SourceContext {
```

also, since the corresponding method is called `with_ctx` (which I'd also rename to `with_context`), we should consider simply calling this `Context`

```suggestion
pub enum Context {
```